### The very first Data-Science project 

![Alt text](https://images.indianexpress.com/2018/10/titanic-759.jpg "a title")

In [1]:
import numpy as np
import pandas as pd

In [2]:
train_df = pd.read_csv('./kaggle/titanic/train.csv')

In [3]:
test_df = pd.read_csv('./kaggle/titanic/test.csv')

In [4]:
train_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

#### Notes
- [SibSp] as number of siblings / spouses aboard the Titanic
- [Parch] as number of parents / children aboard the Titatic
- [Ticket] as ticket number
- [Cabin] as cabin number

In [7]:
train_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [8]:
test_df.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [9]:
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### Nhận xét:
1. Cột [PassengerId] không thực sự hữu ích trong quá trình train nên sẽ loại bỏ


In [35]:
# Đổi cột PassengerId thành index
train_df.set_index(train_df.PassengerId)

PassengerId Survived Pclass  \
PassengerId                                
1                      1        0      3   
2                      2        1      1   
3                      3        1      3   
4                      4        1      1   
5                      5        0      3   
...                  ...      ...    ...   
887                  887        0      2   
888                  888        1      1   
889                  889        0      3   
890                  890        1      1   
891                  891        0      3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

            SibSp Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                        
1               1     0         A/5 21171   7.2500   NaN        S  
2               1     0          PC 17599  71.2833   C85        C  
3               0     0  STON/O2. 3101282   7.9250   NaN        S  
4               1     0            113803  53.1000  C123        S  
5               0     0            373450   8.0500   NaN        S  
...           ...   ...               ...      ...   ...      ...  
887             0     0            211536  13.0000   NaN        S  
888             0     0            112053  30.0000   B42        S  
889             1     2        W./C. 6607  23.4500   NaN        S  
890             0     0            111369  30.0000  C148        C  
891             0     0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [36]:
# Drop cột PassengerId
train_df = train_df.drop('PassengerId', axis = 1)

Survived Pclass                                               Name  \
0          0      3                            Braund, Mr. Owen Harris   
1          1      1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2          1      3                             Heikkinen, Miss. Laina   
3          1      1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4          0      3                           Allen, Mr. William Henry   
..       ...    ...                                                ...   
886        0      2                              Montvila, Rev. Juozas   
887        1      1                       Graham, Miss. Margaret Edith   
888        0      3           Johnston, Miss. Catherine Helen "Carrie"   
889        1      1                              Behr, Mr. Karl Howell   
890        0      3                                Dooley, Mr. Patrick   

        Sex   Age SibSp Parch            Ticket     Fare Cabin Embarked  
0      male  22.0     1     0         A/5 21171   7.2500   NaN        S  
1    female  38.0     1     0          PC 17599  71.2833   C85        C  
2    female  26.0     0     0  STON/O2. 3101282   7.9250   NaN        S  
3    female  35.0     1     0            113803  53.1000  C123        S  
4      male  35.0     0     0            373450   8.0500   NaN        S  
..      ...   ...   ...   ...               ...      ...   ...      ...  
886    male  27.0     0     0            211536  13.0000   NaN        S  
887  female  19.0     0     0            112053  30.0000   B42        S  
888  female   NaN     1     2        W./C. 6607  23.4500   NaN        S  
889    male  26.0     0     0            111369  30.0000  C148        C  
890    male  32.0     0     0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [14]:
# Lúc insert dữ liệu có thể set ngay index column
test_df = pd.read_csv('./kaggle/titanic/test.csv', index_col = 'PassengerId')

In [15]:
test_df

Pclass                                          Name     Sex  \
PassengerId                                                                 
892               3                              Kelly, Mr. James    male   
893               3              Wilkes, Mrs. James (Ellen Needs)  female   
894               2                     Myles, Mr. Thomas Francis    male   
895               3                              Wirz, Mr. Albert    male   
896               3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   
...             ...                                           ...     ...   
1305              3                            Spector, Mr. Woolf    male   
1306              1                  Oliva y Ocana, Dona. Fermina  female   
1307              3                  Saether, Mr. Simon Sivertsen    male   
1308              3                           Ware, Mr. Frederick    male   
1309              3                      Peter, Master. Michael J    male   

              Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
PassengerId                                                                   
892          34.5      0      0              330911    7.8292   NaN        Q  
893          47.0      1      0              363272    7.0000   NaN        S  
894          62.0      0      0              240276    9.6875   NaN        Q  
895          27.0      0      0              315154    8.6625   NaN        S  
896          22.0      1      1             3101298   12.2875   NaN        S  
...           ...    ...    ...                 ...       ...   ...      ...  
1305          NaN      0      0           A.5. 3236    8.0500   NaN        S  
1306         39.0      0      0            PC 17758  108.9000  C105        C  
1307         38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
1308          NaN      0      0              359309    8.0500   NaN        S  
1309          NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 10 columns]

# 1. Feature Classification: Categorical vs Numerical

### Which features are categorical ?
- Categorical feature: nominal, ordinal, ratio, interval


### Which features are numerical ?
- Numerical: discrete - Dữ liệu rời rác, continuous - liên tục, timeseries - thời gian

In [16]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Lưu ý:
Chúng ta có thể thấy cột [Age] và cột [Cabin] không đủ 891 dữ liệu - 

<span style="color:red">`Do đó phải take note để tiền xử lý dữ liệu`</span>

In [17]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 418 entries, 892 to 1309
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    418 non-null    int64  
 1   Name      418 non-null    object 
 2   Sex       418 non-null    object 
 3   Age       332 non-null    float64
 4   SibSp     418 non-null    int64  
 5   Parch     418 non-null    int64  
 6   Ticket    418 non-null    object 
 7   Fare      417 non-null    float64
 8   Cabin     91 non-null     object 
 9   Embarked  418 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 35.9+ KB


### Phân loại features
1. Categorical: `Survived`, `Sex`, `Embarked`, `Pclass` (oridinal), `SibSp`, `Parch`
2. Numeric: `Age` (continuous), `Fare`(discrete)
3. Mix type of data: `Ticket`, `Cabin`
4. Contain Error/Typo
5. Blank or Null: `Cabin` > `Age` > `Embarked`
6. Various Data Type: int, String, Float

### Casting type cho các kiểu dữ liệu về Categorical

- [Survived]

In [18]:
train_df['Survived'] = train_df['Survived'].astype("category")

In [19]:
train_df.Survived.dtype

CategoricalDtype(categories=[0, 1], ordered=False)

In [20]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null    category
 2   Pclass       891 non-null    int64   
 3   Name         891 non-null    object  
 4   Sex          891 non-null    object  
 5   Age          714 non-null    float64 
 6   SibSp        891 non-null    int64   
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float64 
 10  Cabin        204 non-null    object  
 11  Embarked     889 non-null    object  
dtypes: category(1), float64(2), int64(4), object(5)
memory usage: 77.7+ KB


- Categorical: Survived, Sex, Embarked, Pclass (oridinal), SibSp, Parch

In [21]:
categorical_features = ['Pclass','Embarked','Sex','SibSp','Parch','Survived']

In [27]:
def convert_categorical(df, features):
    for feature in features:
        df[feature] = df[feature].astype('category')
        
convert_categorical(train_df, categorical_features)
convert_categorical(test_df, categorical_features)

In [48]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null    category
 2   Pclass       891 non-null    category
 3   Name         891 non-null    object  
 4   Sex          891 non-null    category
 5   Age          714 non-null    float64 
 6   SibSp        891 non-null    category
 7   Parch        891 non-null    category
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float64 
 10  Cabin        204 non-null    object  
 11  Embarked     889 non-null    category
dtypes: category(6), float64(2), int64(1), object(3)
memory usage: 48.3+ KB


In [29]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 418 entries, 892 to 1309
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Pclass    418 non-null    category
 1   Name      418 non-null    object  
 2   Sex       418 non-null    category
 3   Age       332 non-null    float64 
 4   SibSp     418 non-null    category
 5   Parch     418 non-null    category
 6   Ticket    418 non-null    object  
 7   Fare      417 non-null    float64 
 8   Cabin     91 non-null     object  
 9   Embarked  418 non-null    category
dtypes: category(5), float64(2), object(3)
memory usage: 22.7+ KB


# 2. Tìm hiểu phân bố - Distribution of Numerical features values across the samples (trên toàn bộ tập mẫu)

In [54]:
# Phân bố dữ liệu của Numerical features
train_df
train_df.describe()

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

In [55]:
# Phân bố dữ liệu của category features
train_df.describe(include=['category'])

Survived  Pclass   Sex  SibSp  Parch Embarked
count        891     891   891    891    891      889
unique         2       3     2      7      7        3
top            0       3  male      0      0        S
freq         549     491   577    608    678      644




# 3. Exploratory Data Analysis (EDA) - Phân tích khai phá dữ liệu